### ISSUE TITLE
Host won't respond to workload ARPs if host has no default route

### ISSUE THREAD

**Issue by fasaxc** [projectcalico/calico#3270]
For historic reasons (OpenStack) we use proxy ARP to get the host to respond to ARPs for all workloads.  However, a Linux node with proxy ARP turned on will only respond to ARPs for IPs to which is has some route.  This means it doesn't work on air-gapped networks with no default route.

## Expected Behavior

Host should respond to ARPs even if it has no non-Calico routes configured.

## Current Behavior

Host only responds to ARPs if it has a route to 169.245.1.1, the router address.  This requirement is typically satisfied by having a default route.

## Possible Solution

We do the proxy ARP trick for OpenStack, where we can't control the routes inside the workload.  For container-based systems, we could simply add 169.245.1.1 as an IP on the host side of the veth.

## Steps to Reproduce (for bugs)

1. Set up a topology with no default route
2. Create a pod
3. Snoop cali interface of pod, see no ARP responses.

## Your Environment

* Calico version
* Orchestrator version (e.g. kubernetes, mesos, rkt):
* Operating System and version:
* Link to your project (optional):

**Comment by sheepbrew** [comment]
We ran into this issue in our deployment.

Our setup involves installing on bare metal servers that do routing on the host. These servers then connect to ToR (top of rack) switches via BGP. By default, most servers connect to a VRF that has only private routes (RFC1918 and RFC6598), and subsequently only receive these routes.  Servers that need to communicate with the internet connect to another VRF and receive default route from that.

We noticed that proxy ARP was failing on hosts that don't have a default route, which led us to discover this issue.

We worked this around by using @fasaxc suggestion of adding `169.254.1.1/32` to the loopback interface.

Btw, in the original post, `169.245.1.1` should be `169.254.1.1` instead.
> Host only responds to ARPs if it has a route to 169.245.1.1, the router address. This requirement is typically satisfied by having a default route.

**Comment by fasaxc** [comment]
Fixed the comment, thanks.

**Comment by scuzhanglei** [comment]
In my situation, if I delete pod's default route `ip route del default via 169.254.1.1 dev eth0`, the pod eth0 would't response ARPs too.

**Comment by fasaxc** [comment]
@scuzhanglei Yes, if you delete that route things will break; why did you delete that essential route?

**Comment by scuzhanglei** [comment]
@fasaxc I am just curious...  thanks and forget that.

**Comment by zq-david-wang** [comment]
It is really hard to debug this kind of issue. 
Could calico-node  abort when  "169.254.1.1" could not be used?

**Comment by champtar** [comment]
Any reasons we can't set permanent neighbour ?
```
ip neigh replace 169.254.1.1 lladdr ee:ee:ee:ee:ee:ee dev eth0 nud permanent
```
And if we also set the neighbour on the host side we could even switch NOARP on the interfaces

**Comment by champtar** [comment]
BTW adding 169.254.1.1 on the host can cause issue when using kube-proxy strictARP: true, which is required to use MetalLB + kube-proxy IPVS

**Comment by fasaxc** [comment]
Unfortunately, Calico needs to be compatible with non-Calico CNI plugins, so, yes, _we_ could set static ARP but we'd also need all of the other plugins we support to do the same.

We also tried simply assigning the IP on the host side of the veth but it caused unexpected issues in testing.

**Comment by champtar** [comment]
I just tested and adding a route on the host
```
ip r add 169.254.1.1/32 dev lo
```
is working.
When pinging 169.254.1.1 I get
```
root@ubuntu:/# ping 169.254.1.1
PING 169.254.1.1 (169.254.1.1) 56(84) bytes of data.
From 10.10.52.154 icmp_seq=1 Time to live exceeded
From 10.10.52.154 icmp_seq=2 Time to live exceeded
^C
--- 169.254.1.1 ping statistics ---
2 packets transmitted, 0 received, +2 errors, 100% packet loss, time 1045ms
```
But except for that don't know if there are other side effects

**Comment by cmssczy** [comment]
> We also tried simply assigning the IP on the host side of the veth but it caused unexpected issues in testing.

Could you give some details about the `unexpected issues`?

**Comment by fasaxc** [comment]
Haven't tried adding the address to `lo` but adding it to the host side of the veth broke some of our tests.  I don't think we got to the bottom of it.

**Comment by haminhcong** [comment]
I got the same problem when deploy Kubernetes on hosts which don't have default route. Should you add this behvior to Calico install document?

https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises

**Comment by Striar-Yunis** [comment]
The problem is calico is forwarding arp requests via proxy arp. That means the kernel can't tell by the incoming address what the interface of the response needs to be.

K8s Node CIDR: 192.168.0.0/16
K8s Pod CIDR: 10.42.0.0/16
K8s Service CIDR: 10.43.0.0/16

Container1, normal pod:
```
netshoot-567c77c8dd-p7cxm:~# ip r
default via 169.254.1.1 dev eth0 
169.254.1.1 dev eth0 scope link
```

Container2, pod with only routes set (no default):
```
10.42.0.0/16 via 169.254.1.1 dev eth0
10.43.0.0/16 via 169.254.1.1 dev eth0
```

Pinging from Container1 to Container2, logs from `tcpdump -i eth0 -n`:
```
ARP, Request who-has x.x.x.x tell 192.168.1.106 <--- Note that this is the node IP
```
Without a route out the interface it doesn't know where to respond to the arp request.

Adding a route for ARP to the underlying nodes fixes the problem:
```
ip r a 192.168.0.0/16 dev eth0
```

**Comment by fasaxc** [comment]
It'd be good to have a better solution, I came across "proxy" arp table entries; I wonder if we can do something more refined, just add one entry to the arp table for each interface.

**Comment by fasaxc** [comment]
Did a bit of digging today and it seems to be impossible to get the kernel to respond to an ARP for an arbitrary IP unless the kernel has a route to that IP.  The best configuration I could find was the following, but it still requires a route for the Calico IP somewhere:
```
# Strictest settings for ARP filtering:
sysctl -w net.ipv4.conf.veth0.arp_filter=1
sysctl -w net.ipv4.conf.veth0.arp_ignore=2
# Make a route to a dummy interface for the Calico router IP.
ip link add type dummy
ip route add 169.254.1.1 dev dummy0
# Configure ARP proxy just for the IP
ip neigh add proxy 169.254.1.1 dev <veth> extern_learn
```
The alternative is to assign the calico router IP on the host-side of every veth, but when I tried that before it caused some other problems.

**Comment by Striar-Yunis** [comment]
I do wonder if it would make sense to have calico have an additional route table entry though. Calico already adds `169.254.1.1 dev eth0 scope link` which is redundant with `default via 169.254.1.1 dev eth0` unless more routes are added by the user or another CNI. Could have it add an explicit route for the NodeIP the pod is scheduled on? I'm sure that has some hidden implications but would make it clear to Calico users exactly what needs to be routed for CNI functionality.
